In [49]:
import sys
print(sys.executable)

d:\Dev\enterprise-aws-data-lakehouse-ml-system\venv\Scripts\python.exe


In [50]:
import pandas as pd

train_trans = pd.read_csv("../data/raw/train_transaction.csv")
train_id = pd.read_csv("../data/raw/train_identity.csv")

### 1 - Data Understanding

In [51]:
train_trans.head()

,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,...,V330,V331,V332,V333,V334,V335,V336,V337,V338,V339
0,2987000,0,86400,68.5,W,13926,NaN,150.0,discover,142.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2987002,0,86469,59.0,W,4663,490.0,150.0,visa,166.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2987003,0,86499,50.0,W,18132,567.0,150.0,mastercard,117.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2987004,0,86506,50.0,H,4497,514.0,150.0,mastercard,102.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [52]:
train_trans.shape

(590540, 394)

In [53]:
train_trans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 590540 entries, 0 to 590539
Columns: 394 entries, TransactionID to V339
dtypes: float64(376), int64(4), object(14)
memory usage: 1.7+ GB


In [54]:
train_trans.isnull().sum()

TransactionID          0
isFraud                0
TransactionDT          0
TransactionAmt         0
ProductCD              0
                   ...  
V335              508189
V336              508189
V337              508189
V338              508189
V339              508189
Length: 394, dtype: int64

In [55]:
missing_percent = (
    train_trans.isna()
    .mean()
    .mul(100)
    .round(2)
    .to_frame("missing_percent")
    .sort_values("missing_percent", ascending=False)
)

missing_percent.head(10)

,missing_percent
dist2,93.63
D7,93.41
D13,89.51
D14,89.47
D12,89.04
D6,87.61
D9,87.31
D8,87.31
V153,86.12
V140,86.12


In [56]:
train_id.head()

,TransactionID,id_01,id_02,id_03,id_04,id_05,id_06,id_07,id_08,id_09,...,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987004,0.0,70787.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,samsung browser 6.2,32.0,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M
1,2987008,-5.0,98945.0,NaN,NaN,0.0,-5.0,NaN,NaN,NaN,...,mobile safari 11.0,32.0,1334x750,match_status:1,T,F,F,T,mobile,iOS Device
2,2987010,-5.0,191631.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,...,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,Windows
3,2987011,-5.0,221832.0,NaN,NaN,0.0,-6.0,NaN,NaN,NaN,...,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,NaN
4,2987016,0.0,7460.0,0.0,0.0,1.0,0.0,NaN,NaN,0.0,...,chrome 62.0,24.0,1280x800,match_status:2,T,F,T,T,desktop,MacOS


In [57]:
train_id.shape

(144233, 41)

In [58]:
train_id.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144233 entries, 0 to 144232
Data columns (total 41 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionID  144233 non-null  int64  
 1   id_01          144233 non-null  float64
 2   id_02          140872 non-null  float64
 3   id_03          66324 non-null   float64
 4   id_04          66324 non-null   float64
 5   id_05          136865 non-null  float64
 6   id_06          136865 non-null  float64
 7   id_07          5155 non-null    float64
 8   id_08          5155 non-null    float64
 9   id_09          74926 non-null   float64
 10  id_10          74926 non-null   float64
 11  id_11          140978 non-null  float64
 12  id_12          144233 non-null  object 
 13  id_13          127320 non-null  float64
 14  id_14          80044 non-null   float64
 15  id_15          140985 non-null  object 
 16  id_16          129340 non-null  object 
 17  id_17          139369 non-nul

In [59]:
train_id.isnull().sum()

TransactionID         0
id_01                 0
id_02              3361
id_03             77909
id_04             77909
id_05              7368
id_06              7368
id_07            139078
id_08            139078
id_09             69307
id_10             69307
id_11              3255
id_12                 0
id_13             16913
id_14             64189
id_15              3248
id_16             14893
id_17              4864
id_18             99120
id_19              4915
id_20              4972
id_21            139074
id_22            139064
id_23            139064
id_24            139486
id_25            139101
id_26            139070
id_27            139064
id_28              3255
id_29              3255
id_30             66668
id_31              3951
id_32             66647
id_33             70944
id_34             66428
id_35              3248
id_36              3248
id_37              3248
id_38              3248
DeviceType         3423
DeviceInfo        25567
dtype: int64

In [60]:
missing_percent_train_id = (
    train_id.isna()
    .mean()
    .mul(100)
    .round(2)
    .to_frame("missing_percent")
    .sort_values("missing_percent", ascending=False)
)

missing_percent_train_id.head(10)

,missing_percent
id_24,96.71
id_25,96.44
id_07,96.43
id_08,96.43
id_21,96.42
id_22,96.42
id_23,96.42
id_27,96.42
id_26,96.42
id_18,68.72


#### 1.1 Check column existence

In [61]:
print("train_transaction columns contain 'isFraud':",
      "isFraud" in train_trans.columns)

print("train_identity columns contain 'isFraud':",
      "isFraud" in train_id.columns)

train_transaction columns contain 'isFraud': True
train_identity columns contain 'isFraud': False


#### 1.2 Merge Strategy

In [62]:
df = train_trans.merge(train_id, on="TransactionID", how="left")
df.shape

(590540, 434)

## 2 - Missing Data Strategy Design
### 2.1 Missing Value Audit

In [63]:
df.isnull().sum()

TransactionID          0
isFraud                0
TransactionDT          0
TransactionAmt         0
ProductCD              0
                   ...  
id_36             449555
id_37             449555
id_38             449555
DeviceType        449730
DeviceInfo        471874
Length: 434, dtype: int64

In [64]:
df["isFraud"].mean()

np.float64(0.03499000914417313)

In [65]:
# Check if target exists
if "isFraud" in df.columns:
    print("Target column found ✅")
else:
    print("Target column NOT found ❌")

# Class distribution
fraud_counts = df["isFraud"].value_counts()

print("\nFraud Distribution:")
print(fraud_counts)

imbalance_ratio = fraud_counts[0] / fraud_counts[1]
print("Imbalance Ratio:", imbalance_ratio, "\n")

print(df["isFraud"].value_counts(normalize=True))  # normalize=True gives percentage distribution

# Percentage distribution
fraud_percent = df["isFraud"].value_counts(normalize=True) * 100  

print("\nFraud Percentage:")
print(fraud_percent)

Target column found ✅

Fraud Distribution:
isFraud
0    569877
1     20663
Name: count, dtype: int64
Imbalance Ratio: 27.579586700866283 

isFraud
0    0.96501
1    0.03499
Name: proportion, dtype: float64

Fraud Percentage:
isFraud
0    96.500999
1     3.499001
Name: proportion, dtype: float64


### 2.2 Build Missing Summary Table

In [66]:
missing_summary = pd.DataFrame({
    "missing_count": df.isnull().sum(),
    "missing_percent": df.isnull().mean() * 100,
    "dtype": df.dtypes
}).sort_values("missing_percent", ascending=False)

missing_summary

,missing_count,missing_percent,dtype
id_24,585793,99.196159,float64
id_25,585408,99.130965,float64
id_07,585385,99.127070,float64
id_08,585385,99.127070,float64
id_21,585381,99.126393,float64
...,...,...,...
C11,0,0.000000,float64
C14,0,0.000000,float64
C13,0,0.000000,float64
C12,0,0.000000,float64


### 2.3 Test Missing Signal Strength

In [67]:
high_missing_cols = missing_summary[
    missing_summary["missing_percent"] > 99
].index.tolist()

signal_results = {}

for col in high_missing_cols:
    fraud_rate = df.groupby(df[col].isnull())["isFraud"].mean()

    if len(fraud_rate) == 2:  # ensure both True/False exist
        signal_strength = abs(fraud_rate[True] - fraud_rate[False])
        signal_results[col] = signal_strength

# Convert to DataFrame
signal_df = pd.DataFrame.from_dict(
    signal_results,
    orient="index",
    columns=["fraud_rate_difference"]
).sort_values("fraud_rate_difference", ascending=False)

signal_df.head()

,fraud_rate_difference
id_24,0.050098
id_07,0.048068
id_08,0.048068
id_21,0.048003
id_27,0.047843


## fraud_rate[True]

Fraud rate when the column value is **missing (NULL)**:

$$
fraud\_rate[True] =
\frac{Fraud\ Cases\ (col\ is\ NULL)}
{Total\ Rows\ (col\ is\ NULL)}
$$

---

## fraud_rate[False]

Fraud rate when the column value is **NOT missing (NOT NULL)**:

$$
fraud\_rate[False] =
\frac{Fraud\ Cases\ (col\ is\ NOT\ NULL)}
{Total\ Rows\ (col\ is\ NOT\ NULL)}
$$


## Signal Strength Equation

Signal strength measures the absolute difference between fraud rates
when the column is missing and when it is not missing.

$$
signal\_strength = 
\left| 
fraud\_rate[True] - fraud\_rate[False] 
\right|
$$

             A small difference → weak signal
             A large difference → strong signal

### 2.4 Separate Feature Types

In [68]:
target_col = "isFraud"

numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns.tolist()
categorical_cols = df.select_dtypes(include=["object"]).columns.tolist()

# Remove target from numeric list
numeric_cols.remove(target_col)

print("Numeric features:", len(numeric_cols))
print("Categorical features:", len(categorical_cols))

Numeric features: 402
Categorical features: 31


## Key Findings

- Fraud rate ≈ 3.5%
- Imbalance ratio ≈ 27:1
- Several high-missing columns show strong fraud signal
- Dataset has 402 numeric and 31 categorical features
- Time-based validation required